core/translate/index.rs

```diff
@@ -506,29 +523,180 @@ pub fn resolve_sorted_columns(
 ) -> crate::Result<Vec<IndexColumn>> {
     let mut resolved = Vec::with_capacity(cols.len());
     for sc in cols {
-        let ident = match sc.expr.as_ref() {
-            // SQLite supports indexes on arbitrary expressions, but we don't (yet).
-            // See "How to use indexes on expressions" in https://www.sqlite.org/expridx.html
-            Expr::Id(col_name) | Expr::Name(col_name) => col_name.as_str(),
-            _ => crate::bail_parse_error!("Error: cannot use expressions in CREATE INDEX"),
-        };
-        let Some(col) = table.get_column(ident) else {
-            crate::bail_parse_error!(
-                "Error: column '{ident}' does not exist in table '{}'",
-                table.name
-            );
-        };
+        let order = sc.order.unwrap_or(SortOrder::Asc);
+        let (explicit_collation, base_expr) = extract_collation(sc.expr.as_ref())?;
+        if let Some((pos, column_name, column)) = resolve_index_column(base_expr, table) {
+            let collation = explicit_collation.or_else(|| column.collation_opt());
+            resolved.push(IndexColumn {
+                name: column_name,
+                order,
+                pos_in_table: pos,
+                collation,
+                default: column.default.clone(),
+                expr: None,
+            });
+            continue;
+        }
+        if !validate_index_expression(sc.expr.as_ref(), table) {
+            crate::bail_parse_error!("Error: invalid expression in CREATE INDEX: {}", sc.expr);
+        }
         resolved.push(IndexColumn {
-            name: col.1.name.as_ref().unwrap().clone(),
-            order: sc.order.unwrap_or(SortOrder::Asc),
-            pos_in_table: col.0,
-            collation: col.1.collation_opt(),
-            default: col.1.default.clone(),
+            name: sc.expr.to_string(),
+            order,
+            pos_in_table: EXPR_INDEX_SENTINEL,
+            collation: explicit_collation,
+            default: None,
+            expr: Some(sc.expr.clone()),
         });
     }
     Ok(resolved)
 }
 
+/// Extracts collation sequence from an expression if it is a Collate expression.
+/// Given the example: `col1 COLLATE NOCASE` / Expr::Collation(Expr::Id(col1), CollationSeq)
+/// returns (Some(CollationSeq), Expr::Id(col1))
+fn extract_collation(expr: &Expr) -> crate::Result<(Option<CollationSeq>, &Expr)> {
```

Comment:
Can we use existing `get_collseq_from_expr` for this, or is this explicitly only used when the top level expression is `Collate(<expr>)`? Follow-up question: if it's only used for explicit top-level COLLATE, why?